# Lab 01: MỐI QUAN HỆ CỦA DỮ LIỆU
# Môn: Trực Quan Hóa Dữ Liệu

## Danh sách thành viên:

---
**Họ và tên:** Nguyễn Tất Trường

MSSV: 19127082

---
**Họ và tên:** Nguyễn Hữu Tuấn

**MSSV:** 19127083

---
**Họ và tên:** Chung Thế Thọ

**MSSV:** 19127562

---
**Họ và tên:** Trần Quốc Việt

**MSSV:** 19127630

# Import các thư viện cần thiết

In [56]:
# Import thư viện cần dùng
import glob, os
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Đường dẫn đến các thư mục
datapath = 'dataset/'

# Thu thập dữ liệu

Trong đồ án này, ta sẽ sử dụng tập dữ liệu về đại dịch **Coronavirus** thu thập từ trang **Worldometer** (www.worldometers.info/coronavirus/). Dữ liệu được thu thập từ ngày *20-02-2022* đến ngày *10-03-2022* và không liên tục, nghĩa là có vài ngày chưa thu thập dữ liệu. Theo như mô tả trên Worldometer thì dữ liệu của trang này cung cấp cho ta thông tin về sự tác động của **Coronavirus** lên **225 đất nước và vùng lãnh thổ** khác nhau qua từng ngày.

Về cách thức thu thập dữ liệu thì có thể xem chi tiết ở notebook `\dataset\scraping_worldometers.ipynb`

Link của tập dữ liệu đã thu thập: ...

---

# Khám phá dữ liệu

## Đọc dữ liệu của ngày đầu tiên từ file csv vào dataframe

Dữ liệu dạng số lưu trữ trong file được phân cách phần ngàn với nhau bằng dấu phẩy (ví dụ 1,000,000), vì vậy ta truyền vào tham số **thousand=','** khi đọc dữ liệu để cho thư viện pandas tự động chuyển các số này về dạng số thông thường

In [57]:
filename = '2022-2-20.csv'
filepath = datapath + filename
first_day_df = pd.read_csv(filepath, thousands=',')
first_day_df.head(10)

#  Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0 NaN           Asia    112080501     521505       1333376        1766   
1 NaN  North America     94158432      57642       1384321         979   
2 NaN  South America     53321618      91557       1248687         984   
3 NaN         Europe    150719217     660784       1681452        2030   
4 NaN         Africa     11442875       8624        247689         194   
5 NaN        Oceania      3284142      20201          7457          49   
6 NaN            NaN          721        NaN            15         NaN   
7 NaN          World    425007506    1360313       5902997        6002   
8   1          China       107707        195          4636         NaN   
9   2            USA     80117484      28647        959863         431   

   Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0        104076728         498347       6670397              30563   
1         63866758         164163      28907353              18121   
2         46304681         148298       5768250              13954   
3        122959179        1165527      26078586              16595   
4         10454264          15330        740922               1984   
5          2997343          21041        279342                263   
6              706            NaN             0                  0   
7        350659659        2012706      68444850              81480   
8           101449             55          1622                  7   
9         51675064         129104      27482557              11668   

   Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0               NaN            NaN          NaN           NaN         NaN   
1               NaN            NaN          NaN           NaN         NaN   
2               NaN            NaN          NaN           NaN         NaN   
3               NaN            NaN          NaN           NaN         NaN   
4               NaN            NaN          NaN           NaN         NaN   
5               NaN            NaN          NaN           NaN         NaN   
6               NaN            NaN          NaN           NaN         NaN   
7             54525            757          NaN           NaN         NaN   
8                75              3    160000000        111163  1439323776   
9            239741           2872    938796430       2809224   334183570   

              Region  
0               Asia  
1      North America  
2      South America  
3             Europe  
4             Africa  
5  Australia/Oceania  
6                NaN  
7                All  
8               Asia  
9      North America

## Mỗi cột có ý nghĩa gì?

- **#**: Số thứ tự
- **Country,Other**: Tên đất nước/vùng lãnh thổ
- **Total Cases**: Tổng số ca nhiễm
- **New Cases**: Số ca mới trong ngày
- **Total Deaths**: Tổng số ca tử vong
- **New Deaths**: Số ca tử vong trong ngày
- **Total Recovered**: Tổng số ca đã khỏi bệnh
- **New Recovered**: Số ca khỏi bệnh trong ngày
- **Active Cases**: Số ca đang còn bệnh
- **Serious, Critical**: Số ca nghiêm trọng
- **Tot Cases/1M pop**: Tổng ca nhiễm/triệu người
- **Deaths/1M pop**: Tổng ca tử vong/triệu người
- **Total Tests**: Tổng số lần test
- **Tests/1M pop**: Số test/triệu người
- **Population**: Dân số của đất nước/vùng lãnh thổ
- **Region**: Khu vực của đất nước/vùng lãnh thổ

## Số dòng, số cột của tập dữ liệu

In [58]:
num_rows, num_cols = first_day_df.shape
print('Số dòng:', num_rows)
print('Số cột:', num_cols)

Số dòng: 234
Số cột: 16


**Nhận xét:**

Số dòng thực tế nhiều hơn so với số dòng của mô tả trên Worldometers (234 so với 225). Nguyên do của việc này ta có thể thấy ở đầu tập dữ liệu:
- 7 dòng tổng hợp số liệu theo Region/World
- 1 dòng không có ý nghĩa (dòng thứ 7)
Ngoài ra có thể mô tả trên trang Worldometer chưa được cập nhật, bằng chứng là khi ta xem dòng cuối cùng của tập dữ liệu, ta thấy số thứ tự thông qua cột **#** không phải là 225.

In [59]:
first_day_df.tail(1)

# Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
233 226    Micronesia            1        NaN           NaN         NaN   

     Total Recovered  New Recovered  Active Cases  Serious, Critical  \
233                1            NaN             0                NaN   

     Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
233                 9            NaN          NaN           NaN      116999   

                Region  
233  Australia/Oceania

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng trong tập dữ liệu cho ta biết thông tin về tình hình các ca nhiễm, ca khỏi Covid-19 của một đất nước/vùng lãnh thổ cụ thể. Ngoài trừ dòng thứ 7 không mang ý nghĩa mà ta đã đề cập trên, các dòng còn lại có vẻ như không có gì bất ổn. Vì vậy ta sẽ tiến hành xóa đi dòng không có ý nghĩa để thuận tiện cho xử lý sau này.

In [60]:
first_day_df = first_day_df.drop(6)

## Dữ liệu có bị lặp không

In [61]:
print("Duplicated count:", np.sum(first_day_df.duplicated()))

Duplicated count: 0


## Kiểu dữ liệu của từng cột

In [62]:
first_day_df.dtypes

#                    float64
Country,Other         object
Total Cases            int64
New Cases            float64
Total Deaths         float64
New Deaths           float64
Total Recovered      float64
New Recovered        float64
Active Cases         float64
Serious, Critical    float64
Tot Cases/1M pop     float64
Deaths/1M pop        float64
Total Tests          float64
Tests/1M pop         float64
Population           float64
Region                object
dtype: object

**Nhận xét:**

Có vẻ kiểu dữ liệu của các cột không có gì bất ổn. Để đảm bảo ta đi xem các cột mang kiểu *object* có chỉ chứa các dữ liệu dạng chuỗi hay không.

## Phân tích rõ hơn các cột có kiểu object

In [63]:
def open_object_dtype(s):
    return set(s.apply(type))
    
print('Country,Other:', open_object_dtype(first_day_df['Country,Other']))
print('Region:', open_object_dtype(first_day_df['Region']))

Country,Other: {<class 'str'>}
Region: {<class 'str'>, <class 'float'>}


**Nhận xét:**

Kiểu dữ liệu ở cột **Country,Other** đã đúng như ta mong đợi, tuy nhiên ở cột **Region** thì theo kết quả cho ta thấy, có các dòng nào đó mang kiểu dữ liệu float. Ta sẽ xem thử dữ liệu của các dòng này.

In [64]:
first_day_df[first_day_df['Region'].apply(type) == float]

#     Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
219 212  Diamond Princess          712        NaN            13         NaN   
229 222        MS Zaandam            9        NaN             2         NaN   

     Total Recovered  New Recovered  Active Cases  Serious, Critical  \
219              699            NaN             0                NaN   
229                7            NaN             0                NaN   

     Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
219               NaN            NaN          NaN           NaN         NaN   
229               NaN            NaN          NaN           NaN         NaN   

    Region  
219    NaN  
229    NaN

Hai dòng kết quả trả về là dữ liệu của **Diamond Princess** và **MS Zaandam**. Nếu như theo dõi báo đài, tin tức thì ta sẽ biết đây là tên của hai con du thuyền du lịch và vì là du thuyền nên chúng không có **Region** xác định.

## Các cột có bị thiếu dữ liệu không?

In [65]:
first_day_df.isnull().any()

#                     True
Country,Other        False
Total Cases          False
New Cases             True
Total Deaths          True
New Deaths            True
Total Recovered       True
New Recovered         True
Active Cases          True
Serious, Critical     True
Tot Cases/1M pop      True
Deaths/1M pop         True
Total Tests           True
Tests/1M pop          True
Population            True
Region                True
dtype: bool

**Nhận xét:**
    
Chỉ có 2 cột **Country,Other** và **Total Cases** là đầy đủ còn lại tất cả các cột đều bị thiếu dữ liệu.

## Phân bố các cột dữ liệu dạng numeric

Ở mỗi cột ta sẽ xem xét % tỉ lệ thiếu dữ liệu (0 đến 100), giá trị nhỏ nhất và lớn nhất.

In [66]:
pd.options.display.float_format = '{:.0f}'.format

def missing_ratio(s):
    return s.isna().mean() * 100


def min_value(s):
    return s.dropna().min()


def max_value(s):
    return s.dropna().max()


nume_col_profiles_df = first_day_df.select_dtypes(exclude=['object'])


nume_col_profiles_df  = nume_col_profiles_df.agg([missing_ratio,
                                                min_value,
                                                max_value])

nume_col_profiles_df


#  Total Cases  New Cases  Total Deaths  New Deaths  \
missing_ratio   3            0         30             4          49   
min_value       1            1          1             1           1   
max_value     226    425007506    1360313       5902997        6002   

               Total Recovered  New Recovered  Active Cases  \
missing_ratio                4             39             3   
min_value                    1              1             0   
max_value            350659659        2012706      68444850   

               Serious, Critical  Tot Cases/1M pop  Deaths/1M pop  \
missing_ratio                 27                 3              7   
min_value                      1                 9              2   
max_value                  81480            651738           6211   

               Total Tests  Tests/1M pop  Population  
missing_ratio           10            10           4  
min_value              863          5113         804  
max_value        938796430      21356983  1439323776

**Nhận xét:**

Ta nhận thấy dữ liệu ở các cột **New** và **Serious, Critical** bị thiếu khá nhiều. Để lí giải cho điều này thì có lẽ lí do thích hợp nhất là nhiều nước không cập nhật số liệu hằng ngày nên ta không có dữ liệu mà thu thập.

## Phân bố các cột dữ liệu dạng categorical

Ở mỗi cột ta sẽ xem xét % tỉ lệ thiếu dữ liệu (0 đến 100), số giá trị khác nhau (không xét giá trị thiếu) và xem một vài giá trị khác nhau đó.

In [67]:
cate_col_profiles_df = first_day_df.select_dtypes(include=['object'])


def num_diff_vals(s):
    return s.dropna().nunique()

def diff_vals(s):
    return s.dropna().unique()


cate_col_profiles_df = cate_col_profiles_df.agg([missing_ratio,
                                                num_diff_vals,
                                                diff_vals])

cate_col_profiles_df

Country,Other  \
missing_ratio                                                  0   
num_diff_vals                                                233   
diff_vals      [Asia, North America, South America, Europe, A...   

                                                          Region  
missing_ratio                                                  1  
num_diff_vals                                                  7  
diff_vals      [Asia, North America, South America, Europe, A...

**Nhận xét:**

Chỉ có cột **Region** bị thiếu dữ liệu với tỉ lệ rất thấp và ta có thể đoán được dòng nào bị thiếu dữ liệu ở cột này đó là hai chiếc du thuyền ta đã đề cập ở trên.

---

# Tiền xử lý dữ liệu

Thông qua việc khám phá dữ liệu, ta có các vấn đề cần tiền xử lý sau:
- Loại bỏ cột **#**
- Điền dữ liệu bị thiếu ở các cột
- Thêm cột **Date**: ngày thu thập dữ liệu

## Loại bỏ cột #

Ý nghĩa của cột này là số thứ tự mà ta đã có index trong dataframe của pandas nên ta không cần dùng đến cột này nữa để tiết kiệm không gian lưu trữ.

In [24]:
first_day_df = first_day_df.drop('#', axis=1)
first_day_df.head()

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0           Asia    112080501     521505       1333376        1766   
1  North America     94158432      57642       1384321         979   
2  South America     53321618      91557       1248687         984   
3         Europe    150719217     660784       1681452        2030   
4         Africa     11442875       8624        247689         194   

   Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0        104076728         498347       6670397              30563   
1         63866758         164163      28907353              18121   
2         46304681         148298       5768250              13954   
3        122959179        1165527      26078586              16595   
4         10454264          15330        740922               1984   

   Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0               NaN            NaN          NaN           NaN         NaN   
1               NaN            NaN          NaN           NaN         NaN   
2               NaN            NaN          NaN           NaN         NaN   
3               NaN            NaN          NaN           NaN         NaN   
4               NaN            NaN          NaN           NaN         NaN   

          Region  
0           Asia  
1  North America  
2  South America  
3         Europe  
4         Africa

## Điền dữ liệu bị thiếu ở các cột

### Điền giá trị 0

Các cột có thể điền giá trị 0 là: **"New Cases", "Total Deaths", "New Deaths", "Total Recovered", "New Recovered", "Active Cases", "Serious, Critical", "Total Tests"**. Việc điền 0 vào các cột này là có thể chấp nhận được vì có thể một số quốc gia/vùng lãnh thổ không công bố số liệu nữa hoặc chỉ công bố sau một thời gian dài dẫn đến việc ta không có được số liệu cập nhật hằng ngày ở các quốc gia/vùng lãnh thổ này. Vì thế thay vì đoán giá trị của các cột này, tốt nhất ta cho chúng giá trị 0.

In [25]:
fill_zero_cols = ["New Cases", "Total Deaths", "New Deaths", "Total Recovered",
                  "New Recovered", "Active Cases", "Serious, Critical", "Total Tests"]
first_day_df[fill_zero_cols] = first_day_df[fill_zero_cols].fillna(0)

### Điền giá trị thiếu cho cột Population

In [26]:
first_day_df[first_day_df['Population'].isnull()]

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0                Asia    112080501     521505       1333376        1766   
1       North America     94158432      57642       1384321         979   
2       South America     53321618      91557       1248687         984   
3              Europe    150719217     660784       1681452        2030   
4              Africa     11442875       8624        247689         194   
5             Oceania      3284142      20201          7457          49   
7               World    425007506    1360313       5902997        6002   
219  Diamond Princess          712          0            13           0   
229        MS Zaandam            9          0             2           0   

     Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0          104076728         498347       6670397              30563   
1           63866758         164163      28907353              18121   
2           46304681         148298       5768250              13954   
3          122959179        1165527      26078586              16595   
4           10454264          15330        740922               1984   
5            2997343          21041        279342                263   
7          350659659        2012706      68444850              81480   
219              699              0             0                  0   
229                7              0             0                  0   

     Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0                 NaN            NaN            0           NaN         NaN   
1                 NaN            NaN            0           NaN         NaN   
2                 NaN            NaN            0           NaN         NaN   
3                 NaN            NaN            0           NaN         NaN   
4                 NaN            NaN            0           NaN         NaN   
5                 NaN            NaN            0           NaN         NaN   
7               54525            757            0           NaN         NaN   
219               NaN            NaN            0           NaN         NaN   
229               NaN            NaN            0           NaN         NaN   

                Region  
0                 Asia  
1        North America  
2        South America  
3               Europe  
4               Africa  
5    Australia/Oceania  
7                  All  
219                NaN  
229                NaN

Các dòng thiếu dữ liệu ở cột này là các dòng tổng hợp số liệu theo **Region/World** và 2 con du thuyền ta đã đề cập ở trên. Đối với các dòng tổng hợp thì ta sẽ gom nhóm các quốc gia/vùng lãnh thổ theo **Region**, tính tổng theo cột **Population** rồi lấy kết quả để điền, còn đối với 2 chiếc du thuyền thi ta sẽ điền 0.

In [27]:
# Hai chiếc du thuyền
two_cruises = first_day_df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
first_day_df.loc[two_cruises, 'Population'] = 0

# Gôm nhóm theo region rồi tỉnh tổng theo cột Population
region_pop = first_day_df.groupby('Region')['Population'].sum()
region_pop['All'] = region_pop.sum() # World

# Định nghĩa hàm trả Population khi truyền vào region
def region_population(region_pop, region):
    return region_pop[region]

# Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
first_day_df.loc[0:7, 'Population'] = first_day_df.head(7).apply(lambda row: region_population(region_pop, row['Region']), axis=1)

first_day_df.head(10)

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0            Asia    112080501     521505       1333376        1766   
1   North America     94158432      57642       1384321         979   
2   South America     53321618      91557       1248687         984   
3          Europe    150719217     660784       1681452        2030   
4          Africa     11442875       8624        247689         194   
5         Oceania      3284142      20201          7457          49   
7           World    425007506    1360313       5902997        6002   
8           China       107707        195          4636           0   
9             USA     80117484      28647        959863         431   
10          India     42838524      16051        512141         206   

    Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0         104076728         498347       6670397              30563   
1          63866758         164163      28907353              18121   
2          46304681         148298       5768250              13954   
3         122959179        1165527      26078586              16595   
4          10454264          15330        740922               1984   
5           2997343          21041        279342                263   
7         350659659        2012706      68444850              81480   
8            101449             55          1622                  7   
9          51675064         129104      27482557              11668   
10         42124284          37901        202099               8944   

    Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0                NaN            NaN            0           NaN  4664922717   
1                NaN            NaN            0           NaN   596673758   
2                NaN            NaN            0           NaN   436583945   
3                NaN            NaN            0           NaN   748370013   
4                NaN            NaN            0           NaN  1393144288   
5                NaN            NaN            0           NaN    43379587   
7              54525            757            0           NaN  7883074308   
8                 75              3    160000000        111163  1439323776   
9             239741           2872    938796430       2809224   334183570   
10             30550            365    760146333        542096  1402235728   

               Region  
0                Asia  
1       North America  
2       South America  
3              Europe  
4              Africa  
5   Australia/Oceania  
7                 All  
8                Asia  
9       North America  
10               Asia

### Điền lại giá trị cột Total Tests cho các dòng tổng hợp theo Region/World

Tương tự ta cũng gom nhóm các quốc gia/vùng lãnh thổ theo **Region**, tính tổng theo cột **Total Tests** rồi lấy kết quả để điền.

In [28]:
# Gôm nhóm theo region rồi tỉnh tổng theo cột Total Tests
total_tests_pop = first_day_df.groupby('Region')['Total Tests'].sum()
total_tests_pop['All'] = total_tests_pop.sum() # World

# Định nghĩa hàm trả Total Tests khi truyền vào region
def region_total_tests(total_tests_pop, region):
    return total_tests_pop[region]

# Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
first_day_df.loc[0:7, 'Total Tests'] = first_day_df.head(7).apply(lambda row: region_total_tests(total_tests_pop, row['Region']), axis=1)

first_day_df.head(10)

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0            Asia    112080501     521505       1333376        1766   
1   North America     94158432      57642       1384321         979   
2   South America     53321618      91557       1248687         984   
3          Europe    150719217     660784       1681452        2030   
4          Africa     11442875       8624        247689         194   
5         Oceania      3284142      20201          7457          49   
7           World    425007506    1360313       5902997        6002   
8           China       107707        195          4636           0   
9             USA     80117484      28647        959863         431   
10          India     42838524      16051        512141         206   

    Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0         104076728         498347       6670397              30563   
1          63866758         164163      28907353              18121   
2          46304681         148298       5768250              13954   
3         122959179        1165527      26078586              16595   
4          10454264          15330        740922               1984   
5           2997343          21041        279342                263   
7         350659659        2012706      68444850              81480   
8            101449             55          1622                  7   
9          51675064         129104      27482557              11668   
10         42124284          37901        202099               8944   

    Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0                NaN            NaN   1907487673           NaN  4664922717   
1                NaN            NaN   1050910315           NaN   596673758   
2                NaN            NaN    207805821           NaN   436583945   
3                NaN            NaN   2462184094           NaN   748370013   
4                NaN            NaN     95174825           NaN  1393144288   
5                NaN            NaN     70290543           NaN    43379587   
7              54525            757   5793853271           NaN  7883074308   
8                 75              3    160000000        111163  1439323776   
9             239741           2872    938796430       2809224   334183570   
10             30550            365    760146333        542096  1402235728   

               Region  
0                Asia  
1       North America  
2       South America  
3              Europe  
4              Africa  
5   Australia/Oceania  
7                 All  
8                Asia  
9       North America  
10               Asia

### Điền các cột có dạng Value/1M pop

Ta có 3 cột theo dạng trên 1 triệu dân số: **Tot Cases/1M pop, Deaths/1M pop** và **Tests/1M pop**. Để điền giá trị thiếu cho các cột này ta đơn giản lấy cột tương ứng chia cho cột **Population** rồi nhân cho 1 triệu.

In [29]:
def col_on_1millpop(col, pop):
    if pop == 0: # Không có dân số không thể tính toán
        return 0
    return col/pop*10**6

first_day_df['Tot Cases/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Cases'], row['Population'])
                                                             if pd.isnull(row['Tot Cases/1M pop']) else row['Tot Cases/1M pop'],
                                                     axis=1)

first_day_df['Deaths/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Deaths'], row['Population'])
                                                             if pd.isnull(row['Deaths/1M pop']) else row['Deaths/1M pop'],
                                                     axis=1)

first_day_df['Tests/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Tests'], row['Population'])
                                                             if pd.isnull(row['Tests/1M pop']) else row['Tests/1M pop'],
                                                     axis=1)

### Điền Region cho 2 chiếc du thuyền

Phần Region của 2 chiếc du thuyền này ta sẽ điền chuỗi rỗng.

In [30]:
two_cruises = first_day_df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
first_day_df.loc[two_cruises, 'Region'] = ''

### Kiểm tra lại còn cột nào thiếu dữ liệu không?

In [31]:
first_day_df.isnull().any()

Country,Other        False
Total Cases          False
New Cases            False
Total Deaths         False
New Deaths           False
Total Recovered      False
New Recovered        False
Active Cases         False
Serious, Critical    False
Tot Cases/1M pop     False
Deaths/1M pop        False
Total Tests          False
Tests/1M pop         False
Population           False
Region               False
dtype: bool

## Thêm cột Date: ngày thu thập dữ liệu

In [32]:
first_day_df['Date'] = pd.to_datetime(filename[:-4])
first_day_df.head()

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0           Asia    112080501     521505       1333376        1766   
1  North America     94158432      57642       1384321         979   
2  South America     53321618      91557       1248687         984   
3         Europe    150719217     660784       1681452        2030   
4         Africa     11442875       8624        247689         194   

   Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0        104076728         498347       6670397              30563   
1         63866758         164163      28907353              18121   
2         46304681         148298       5768250              13954   
3        122959179        1165527      26078586              16595   
4         10454264          15330        740922               1984   

   Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0             24026            286   1907487673        408900  4664922717   
1            157806           2320   1050910315       1761281   596673758   
2            122134           2860    207805821        475981   436583945   
3            201397           2247   2462184094       3290062   748370013   
4              8214            178     95174825         68317  1393144288   

          Region       Date  
0           Asia 2022-02-20  
1  North America 2022-02-20  
2  South America 2022-02-20  
3         Europe 2022-02-20  
4         Africa 2022-02-20

## Tiến hành tiền xử lý tất cả các tập dữ liệu

Các bước tiến hành như sau
- **Bước 1:** Liệt kê tất cả các file csv
- **Bước 2:** Xử lý trên từng file rồi export ra file csv khác
- **Bước 3:** Gộp tất cả các file csv đã xử lý thành một file

In [33]:
all_files = glob.glob(os.path.join(datapath, "*.csv"))
all_files = [file[8:] for file in all_files]
all_files

['2022-2-20.csv',
 '2022-2-22.csv',
 '2022-2-23.csv',
 '2022-2-24.csv',
 '2022-2-25.csv',
 '2022-2-26.csv',
 '2022-2-27.csv',
 '2022-3-10.csv',
 '2022-3-2.csv',
 '2022-3-4.csv',
 '2022-3-5.csv',
 '2022-3-6.csv',
 '2022-3-7.csv',
 '2022-3-8.csv',
 '2022-3-9.csv']

In [34]:
for file_name in all_files:  
    filepath = datapath + filename
    df = pd.read_csv(filepath, index_col=False, thousands=',')
    
    # Bỏ dòng không có ý nghĩa
    df = df.drop(df[df['Country,Other'].isnull()].index)
    
    # Bỏ cột #
    df = df.drop('#', axis=1)
    
    fill_zero_cols = ["New Cases", "Total Deaths", "New Deaths", "Total Recovered",
                  "New Recovered", "Active Cases", "Serious, Critical", "Total Tests"]
    df[fill_zero_cols] = df[fill_zero_cols].fillna(0)
    
    # Điền Population cho 2 du thuyền
    two_cruises = df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
    df.loc[two_cruises, 'Population'] = 0

    # Gôm nhóm theo region rồi tỉnh tổng theo cột Population
    region_pop = df.groupby('Region')['Population'].sum()
    region_pop['All'] = region_pop.sum() # World

    # Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
    df.loc[0:7, 'Population'] = df.head(7).apply(lambda row: region_population(region_pop, row['Region']), axis=1)

    # Gôm nhóm theo region rồi tỉnh tổng theo cột Total Tests
    total_tests_pop = df.groupby('Region')['Total Tests'].sum()
    total_tests_pop['All'] = total_tests_pop.sum() # World

    # Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
    df.loc[0:7, 'Total Tests'] = df.head(7).apply(lambda row: region_total_tests(total_tests_pop, row['Region']), axis=1)

    def col_on_1millpop(col, pop):
        if pop == 0: # Không có dân số không thể tính toán
            return 0
        return col/pop*10**6

    # Điền giá trị các cột dạng Value/1M pop
    df['Tot Cases/1M pop'] = df.apply(lambda row: col_on_1millpop(row['Total Cases'], row['Population'])
                                                                 if pd.isnull(row['Tot Cases/1M pop']) else row['Tot Cases/1M pop'],
                                                         axis=1)

    df['Deaths/1M pop'] = df.apply(lambda row: col_on_1millpop(row['Total Deaths'], row['Population'])
                                                                 if pd.isnull(row['Deaths/1M pop']) else row['Deaths/1M pop'],
                                                         axis=1)

    df['Tests/1M pop'] = df.apply(lambda row: col_on_1millpop(row['Total Tests'], row['Population'])
                                                                 if pd.isnull(row['Tests/1M pop']) else row['Tests/1M pop'],
                                                         axis=1)
    
    # Điền region cho 2 du thuyền
    two_cruises = df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
    df.loc[two_cruises, 'Region'] = ''
    
    # Thêm cột Date
    df['Date'] = pd.to_datetime(file_name[:-4])
    
    df.to_csv(datapath + f'clean-{file_name}', index=False)

In [35]:
csv_files = glob.glob(os.path.join(datapath, "*.csv"))
csv_files = [file[8:] for file in csv_files]
csv_files = csv_files[int(len(csv_files)/2):]

columns = ['Country,Other', 'Total Cases', 'New Cases', 'Total Deaths',
       'New Deaths', 'Total Recovered', 'New Recovered', 'Active Cases',
       'Serious, Critical', 'Tot Cases/1M pop', 'Deaths/1M pop', 'Total Tests',
       'Tests/1M pop', 'Population', 'Region', 'Date']

merge = pd.DataFrame(columns=columns)


for file in csv_files:
    temp = pd.read_csv(datapath+file, index_col=False)
    merge = merge.append(temp)
    
merge.to_csv(datapath+'dataset.csv', index=False)

## Đọc tập dữ liệu sau khi đã tổng hợp

In [70]:
df = pd.read_csv(datapath+'dataset.csv', parse_dates=['Date']) # Chỉ ra cột 'Date' có kiểu dữ liệu datetime
df

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0                 Asia    112080501     521505       1333376        1766   
1        North America     94158432      57642       1384321         979   
2        South America     53321618      91557       1248687         984   
3               Europe    150719217     660784       1681452        2030   
4               Africa     11442875       8624        247689         194   
...                ...          ...        ...           ...         ...   
3490        MS Zaandam            9          0             2           0   
3491  Marshall Islands            7          0             0           0   
3492      Cook Islands            4          0             0           0   
3493      Saint Helena            2          0             0           0   
3494        Micronesia            1          0             0           0   

      Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0           104076728         498347       6670397              30563   
1            63866758         164163      28907353              18121   
2            46304681         148298       5768250              13954   
3           122959179        1165527      26078586              16595   
4            10454264          15330        740922               1984   
...               ...            ...           ...                ...   
3490                7              0             0                  0   
3491                7              0             0                  0   
3492                0              0             4                  0   
3493                2              0             0                  0   
3494                1              0             0                  0   

      Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0                24026            286   1907487673        408900  4664922717   
1               157806           2320   1050910315       1761281   596673758   
2               122134           2860    207805821        475981   436583945   
3               201397           2247   2462184094       3290062   748370013   
4                 8214            178     95174825         68317  1393144288   
...                ...            ...          ...           ...         ...   
3490                 0              0            0             0           0   
3491               117              0            0             0       59845   
3492               227              0          863         49062       17590   
3493               327              0            0             0        6107   
3494                 9              0            0             0      116999   

                 Region       Date  
0                  Asia 2022-02-20  
1         North America 2022-02-20  
2         South America 2022-02-20  
3                Europe 2022-02-20  
4                Africa 2022-02-20  
...                 ...        ...  
3490                NaN 2022-03-09  
3491  Australia/Oceania 2022-03-09  
3492  Australia/Oceania 2022-03-09  
3493             Africa 2022-03-09  
3494  Australia/Oceania 2022-03-09  

[3495 rows x 16 columns]